### DISCUSSION TITLE
Component diasable

### DISCUSSION THREAD

**Discussion by jojogib** [RomanKubiak/ctrlr#389] [Q&A]
Hello everyone!
I have a tab with multible modulators inside. When I choose a tab i want to send a sysex string only by changing the tab like this:

```
tab2 = function(mod, tabIndex)
rev = {0xF0, 0x43, 0x11, 0x4C, 0x02, 0x01, 0x40, 0x01, 0x00, 0xF7}
rev2 = {0xF0, 0x43, 0x11, 0x4C, 0x02, 0x01, 0x40, 0x02, 0x00, 0xF7}
rev3 = {0xF0, 0x43, 0x11, 0x4C, 0x02, 0x01, 0x40, 0x03, 0x00, 0xF7}

if tabIndex==0 then
panel:sendMidiMessageNow(CtrlrMidiMessage(rev))

elseif tabIndex==1 then
panel:sendMidiMessageNow(CtrlrMidiMessage(rev2))

elseif tabIndex==2 then
panel:sendMidiMessageNow(CtrlrMidiMessage(rev3))

end
end
```

**But I also want to disable some modulators when I change the tab like (componentDisabled="1"). I tried some methodes but wasnt able to do this. I always got some errors. May someone could help me out. I´m not good in lua.
Thanks for your help**

**Comment by Ted-juh** [comment]
Change the last part into: ("componentDisabled", 1, false)

**Reply by jojogib** [reply]
thank you for your replay, but i dont get it. I dont want to change anything on this code i want to add the disable funktion like

```

tab2 = function(mod, tabIndex)

rev = {0xF0, 0x43, 0x11, 0x4C, 0x02, 0x01, 0x40, 0x01, 0x00, 0xF7}
rev2 = {0xF0, 0x43, 0x11, 0x4C, 0x02, 0x01, 0x40, 0x02, 0x00, 0xF7}
rev3 = {0xF0, 0x43, 0x11, 0x4C, 0x02, 0x01, 0x40, 0x03, 0x00, 0xF7}

if tabIndex==0 then
panel:sendMidiMessageNow(CtrlrMidiMessage(rev))
panel:getModulatorByName("timerev"):getComponent("componentDisabled", 1, false)



elseif tabIndex==1 then
panel:sendMidiMessageNow(CtrlrMidiMessage(rev2))

elseif tabIndex==2 then
panel:sendMidiMessageNow(CtrlrMidiMessage(rev3))

end
end

```
For instance: if tab 1 is called send the sysexstring (this one works), but at the same time disable certain modulators (in this case "timerev")
I hope it gets more clear now. Thanks for yor help.

**Reply by Ted-juh** [reply]
panel:getModulatorByName("somemodulator"):getComponent(): setProperty ("componentDisabled", 1, false)

Can't remember if the 1 should be a 1 or an 0.

**Comment by dobo365** [comment]
Or even better:
panel:getModulatorByName("somemodulator"):getComponent():setEnabled(false)

As I also said several times, it is better to assign all your modulators to variables at panel init than calling all the time the panel:getModulatorByName() function.
So in that method called for example "AssignModulators" you would set all your statements like:
modVCO1pitch = panel:getModulatorByName("VCO1pitch")
modVCO2pitch = panel:getModulatorByName("VCO2pitch")
and so on
In your methods you then just use
modVCO1pitch:getComponent():setEnabled(false)

**Comment by jojogib** [comment]
thank you both for your help. I figuard out to send the right sysex by changing the tab and disable the modulators i want. First step made :) One major mistake I did, I wasn´t aware that it makes a differnce to write ModulatorByName("revtime") or  ModulatorByName("RevTime").

One more thing....

I also need a method to send a "snapshot(current value)" to the modulator who aren´t disabled. Like:
My Modulator(reverb) has this sysex (F0 43 11 4C 02 01 42 MS LS F7)

```
if tabIndex==1 then
reverb = getModulatorByName("reverb"):getModulatorValue()
panel:sendMidiMessageNow(0xf0, 0x43, 0x11, 0x02, 0x01, 0x42, reverb, 0xf7 )
```
I know thats wrong but you might get the idea. 
May you can help with that too. By the way i give my panels away for free so every one benefit. So thank you both for your participation

**Reply by dnaldoog** [reply]
Hi @jojogib,

This is how I split a 14 bit value into MSB and LSB in lua.

First create a 7bit nibbleizing function - Lua can return two values from a function.

<pre>
function create7n(n) -- convert integer into two 7 bit nibbles
    local bv = BigInteger(tonumber(n))
    return bv:getBitRangeAsInt(7, 14), bv:getBitRangeAsInt(0,7)
end
</pre>

So you would create two local variables for MSB and LSB: We'll call them MS,LS:

<pre>
if tabIndex==1 then
local reverb = panel:getModulatorByName("reverb"):getModulatorValue()
local MS,LS=create7n(reverb)
local message=string.format("F0 43 11 02 01 42 %.2X %.2X F7",MS,LS)
 panel:sendMidiMessageNow(CtrlrMidiMessage(message))
end -- tabIndex=1
</pre>

**Reply by jojogib** [reply]
Hi,
thanks it worked only in parts...
When ctrlr is running and I´m compiling my lua script it worked like it should, sending multiple sysex strings with current value. But as soon i save the panel and reopen it I get following message:

Callback error: tab 
At line [36]:[string tab] : attempt to index a nil value.

I think it has to do with this funktion:

```
function create7n(n) -- convert integer into two 7 bit nibbles
    local bv = BigInteger(tonumber(n))
    return bv:getBitRangeAsInt(7, 14), bv:getBitRangeAsInt(0,7)
end
```
Because it works, no metter if I delete it or not. 
So where I´m putting this function?
```

function create7n(n) -- convert integer into two 7 bit nibbles
    local bv = BigInteger(tonumber(n))
    return bv:getBitRangeAsInt(7, 14), bv:getBitRangeAsInt(0,7)
end

tab = function(mod, tabIndex)

rev = {0xF0, 0x43, 0x11, 0x4C, 0x02, 0x01, 0x40, 0x01, 0x00, 0xF7}

if tabIndex==0 then

panel:sendMidiMessageNow(CtrlrMidiMessage(rev))

local TimeRev = panel:getModulatorByName("TimeRev"):getModulatorValue()
local MS,LS=create7n(TimeRev)
local message=string.format("F0 43 11 4C 02 01 42 %.2X %.2X F7",MS,LS)
 panel:sendMidiMessageNow(CtrlrMidiMessage(message))

end
```
Is this correct?

**Comment by jojogib** [comment]
I think I found the problem but not the solution.
If the method is in "called when tab change" the panel doesn´t have the values of the modulators on startup.
But if I put in this method wehn the panel is already loaded it works flawless.
How can I say that this method only works if all modulators are loaded`?

**Comment by dnaldoog** [comment]
The solution to this is not easy, because  any function callback attached to a modulator here "_Called when the modulator value changes_" will be triggered on load, so the solution is to add some code that blocks those functions from running when the panel loads.

In the panel below I set a global variable <code>block=true</code> in a function that loads before any modulators are created (a kind of constructor function). So at the head of any function I add the code <code>if block == true then return end</code> - now those functions are blocked from running, but forever!!!! What to do?

Next we need to create some code that will run after everything has loaded to free up those functions by changing the variable <code>block=false</code>. To do that we have a hidden button that will its callback function at startup i.e. it __doesn't__ have <code>if block == true then return end</code> at the head of the function attached to it. The function attached to it runs a 'timer'  or code that is run only after a period of time lapses. This function frees up all functions <code>block=false</code> but by this time the panel has already loaded so none of these functions in "_Called when the modulator value changes_" will be triggered.

<pre>
tab = function(mod, tabIndex)
if block == true then return end
    local rev = {0xF0, 0x43, 0x11, 0x4C, 0x02, 0x01, 0x40, 0x01, 0x00, 0xF7}
    if tabIndex == 0 then
        panel:sendMidiMessageNow(CtrlrMidiMessage(rev))
        local TimeRev = panel:getModulatorByName("TimeRev"):getModulatorValue()
        local MS, LS = create7n(TimeRev)
        local message = string.format("F0 43 11 4C 02 01 42 %.2X %.2X F7", MS, LS)
        panel:sendMidiMessageNow(CtrlrMidiMessage(message))
    end
end
</pre>

Please have a look at the attached panel - it's a kind of template for how to fix this problem.
Also use Ctrlr version 5.3.201 for Windows or Ctrlr 5.3.198 for macOS.

[block function running.zip](https://github.com/RomanKubiak/ctrlr/files/7689731/block.function.running.zip)

**Reply by jojogib** [reply]
Thank you, got it to run..
Last question for now (hopefully)

```
if tabIndex == 0 then
panel:sendMidiMessageNow(CtrlrMidiMessage(rev))
        local TimeRev = panel:getModulatorByName("TimeRev"):getModulatorValue()
        local MS, LS = create7n(TimeRev)
        local message = string.format("F0 43 11 4C 02 01 42 %.2X %.2X F7", MS, LS)
        panel:sendMidiMessageNow(CtrlrMidiMessage(message))
    end
```
I have some listboxes with listboxcontent (reverb1 = 127, reverb2= 255 ......). These boxes wont work with the method above only the sliders work. What do I have to change ?

**Reply by dnaldoog** [reply]
hi @jojogib,

Do you mean if you change uiTab you want to look up the selected value in a list box and send either 127 or 255 for example?

**Reply by jojogib** [reply]
exactly

**Comment by dnaldoog** [comment]
There's a DEMO panel the developer Atom made of ListBox in which looping through the elements is needed when multiple selection is enabled. It should be in your install directory of Ctrlr. But in your case I think a simple <code>panel:getModulatorByName()</code> works as you are only able to click on one selection. 

Here is a suggested function. There is some code there to stop the function from firing if nether of the two Reverbs are selected for some reason. There are many ways to check for this. Maybe even easier ways!

<pre>
sendReverbFromListBox = function(--[[ CtrlrModulator --]] mod --[[ number --]], tabIndex)
    if block == true then
        return
    end
    if tabIndex == 0 then
        local t = {127, 255}
        local rev = panel:getModulatorByName("myListBox"):getModulatorValue() -- listBox selection 0,1
        local MS, LS = create7n(t[rev + 1] or 0xffff) -- lua tables start at 1 so increment
 		if MS == 511 then return end 
                -- exit if neither selection was detected
		-- if the function was fired but there was no selection on the list box
		-- 0xffff is an arbitrarily large number
        local message = string.format("F0 43 11 4C 02 01 42 %.2X %.2X F7", MS, LS)
        panel:sendMidiMessageNow(CtrlrMidiMessage(message))
        panel:getLabel("debugL"):setText(message)
    end
end
</pre>

[list box reverb selection send_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/7692485/list.box.reverb.selection.send_1_0.zip)
